### PR TITLE
Add new param syntax to evaluate against the server

### PR DESCRIPTION
**Pull request by pe4cey**
## Summary

Introduces additional `:param` syntax that evaluates the expression against neo4j server

## More info

The statement to the right of `=>` will be evaluated against neo4j server (as cypher) and the resulting value stored against the param name to be used in future cypher statements (e.g. RETURN `$name`). 

NOTE: Setting params using `:param foo: "bar"`is still supported  but will _not_ be executed against the neo4j server

#### Browser command - return value _(type)_
`:param foo => "bar"` ->`"bar"` _(string)_
`:param foo => 1 + 1` ->`2` _(bolt int)_
`:param foo => date()` -> `"2018-04-23"` _(bolt date string)_

Fixes https://github.com/neo4j/neo4j-browser/issues/705

### CODE REVIEW

File: src/shared/modules/commands/commandsDuck.test.js [hunk 1]
```diff
@@ -137,6 +144,12 @@ describe('commandsDuck', () => {
       const cmdString = cmd + ' x: 2'
       const id = 1
       const action = commands.executeCommand(cmdString, id)
+      bolt.routedWriteTransaction = jest.fn(() =>
```
Comment (oskarhane) [inline src/shared/modules/commands/commandsDuck.test.js]: This is needed here, because this should not call that function.

File: src/shared/modules/commands/commandsDuck.test.js [hunk 2]
```diff
@@ -175,6 +189,7 @@ describe('commandsDuck', () => {
           { type: commands.KNOWN_COMMAND },
           replaceParams({ x: 2, y: 3 }),
           frames.add({ ...action, success: true, type: 'params', params: {} }),
+          { type: 'meta/FORCE_FETCH' },
```
Comment (oskarhane) [inline src/shared/modules/commands/commandsDuck.test.js]: We shouldn't force fetch on param setting

File: src/shared/modules/commands/helpers/params.js [hunk 3]
```diff
@@ -42,14 +72,33 @@ export const handleParamsCommand = (action, cmdchar, put, store) => {
...
+            const result = recursivelyTypeGraphItems(obj)
+            put(update(result))
+            resolve({ result, type: 'param' })
+          })
+          .catch(e => reject(e))
       } catch (e) {
-        return reject(
-          new Error('Could not parse input. Usage: `:param "x": 2`. ' + e)
-        )
+        reject(new Error('Could not parse input. Usage: `:param "x": 2`. ' + e))
```
Comment (oskarhane) [inline src/shared/modules/commands/helpers/params.js]: I think we should promote the new format here

File: src/shared/modules/commands/helpers/params.test.js [hunk 4]
```diff
@@ -44,6 +50,8 @@ describe('commandsDuck params helper', () => {
   })
   test('handles :param "x": 2 and calls the update action creator', () => {
     // Given
+    // jest.spyOn(params, , accessType?)
```
Comment (oskarhane) [inline src/shared/modules/commands/helpers/params.test.js]: not needed

File: src/shared/services/commandInterpreterHelper.js [hunk 5]
```diff
@@ -90,7 +91,7 @@ const availableCommands = [
     name: 'set-params',
     match: cmd => /^params?\s/.test(cmd),
     exec: function (action, cmdchar, put, store) {
-      handleParamsCommand(action, cmdchar, put)
+      return handleParamsCommand(action, cmdchar, put, bolt)
```
Comment (oskarhane) [inline src/shared/services/commandInterpreterHelper.js]: It looks like `handleParamsCommand ` isn't using the '`bolt` argument so it can be skipped here (and also the import up top)

File: e2e_tests/integration/index.spec.js [hunk 6]
```diff
@@ -108,6 +108,25 @@ describe('Neo4j Browser', () => {
       .get('[data-test-id="sidebarMetaItem"]', { timeout: 30000 })
       .should('have.length', 18)
   })
+  it('will add parameter using `:param` command', () => {
+    // add cypher evalutated param function
+    const command = ':param foo => 1 + 1'
+    cy.get(Editor).type(command, { force: true })
```
Comment (oskarhane) [inline e2e_tests/integration/index.spec.js]: Nice with e2e test! 
You could use `cy.executeCommand(command)` here instead to clean up a bit.

File: src/browser/modules/Stream/ParamsFrame.jsx [hunk 7]
```diff
@@ -28,7 +28,9 @@ const ParamsFrame = ({ frame, params }) => {
   const contents = (
     <PaddedDiv>
       <Render if={frame.success !== false}>
-        <pre>{JSON.stringify(frame.params, null, 2)}</pre>
+        <pre data-test-id='rawParamData'>
+          {JSON.stringify(frame.params, null, 2)}
```
Comment (oskarhane) [inline src/browser/modules/Stream/ParamsFrame.jsx]: I think we want a nicer output for the new types. Might be in a separate PR, let's chat about it.

File: src/shared/modules/commands/commandsDuck.test.js [hunk 8]
```diff
@@ -43,6 +43,10 @@ import {
   replace as replaceSettings
 } from 'shared/modules/settings/settingsDuck'
 import { cleanCommand, getInterpreter } from 'services/commandUtils'
+import bolt from 'services/bolt/bolt'
```
Comment (oskarhane) [inline src/shared/modules/commands/commandsDuck.test.js]: Bolt should not be needed in this file since we're not testing the new format, right?

File: src/shared/modules/commands/helpers/params.js (hunk 3 above)
Comment (pe4cey) [reply src/shared/modules/commands/helpers/params.js]: Display both usages or only the new `=>` approach?

**Comment by pe4cey** [comment]
PR updated @oskarhane

File: src/shared/modules/commands/commandsDuck.test.js (hunk 8 above)
Comment (pe4cey) [reply src/shared/modules/commands/commandsDuck.test.js]: I've just added a test in the latest commit that requires this.

**Comment by oskarhane** [comment]
In https://github.com/neo4j/neo4j-browser/pull/735 I've added a new function that formats all types in a nice way. It'd be great if that could be used when printing the params as well.

File: src/shared/modules/commands/commandsDuck.test.js [hunk 9]
```diff
@@ -137,6 +144,43 @@ describe('commandsDuck', () => {
...
+      // When
+      store.dispatch(action)
+
+      // Then
+      // See above
+    })
+    test('does the right thing for :param x => 2', done => {
+      // Given
+      const cmd = store.getState().settings.cmdchar + 'param'
+      const cmdString = cmd + ' x: 2'
```
Comment (oskarhane) [inline src/shared/modules/commands/commandsDuck.test.js]: This doesn't test what the test title says it should test.

File: src/shared/modules/commands/helpers/params.js (hunk 3 above)
Comment (oskarhane) [reply src/shared/modules/commands/helpers/params.js]: Both, I think.
